### PR TITLE
Address book encryption

### PR DESCRIPTION
**Pull request by str4d**
Blocked on changes to the SDK to expose ZIP 32 Arbitrary key derivation.

# Author

- [ ] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [ ] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [ ] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 1]
```diff
@@ -185,20 +199,36 @@ extension AddressBookClient: DependencyKey {
...
+        let salt = SymmetricKey(size: SymmetricKeySize.bits256)
+        return try salt.withUnsafeBytes { salt in
+            let salt = Data(salt)
+            let subKey = deriveEncryptionKey(addressBookKey: encryptionKey, salt: salt)
+
+            // Encrypt the serialized address book.
+            // CryptoKit encodes the SealedBox as `nonce || ciphertext || tag`.
+            let sealed = try ChaChaPoly.seal(data, using: subKey)
+
+            // Prepand the salt to the SealedBox so we can re-derive the sub-key.
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:212]:
```suggestion
            // Prepend the salt to the SealedBox so we can re-derive the sub-key.
```

File: modules/Sources/Models/AddressBookEncryptionKeys.swift [hunk 2]
```diff
@@ -6,18 +6,27 @@
...
+    public init(key: SymmetricKey?) {
         self.key = key
     }
+    
+    public init(from decoder: any Decoder) throws {
+        self.key = nil
+    }
+    
+    public func encode(to encoder: any Encoder) throws {
+
```
Comment (daira) [inline modules/Sources/Models/AddressBookEncryptionKeys.swift:24]:
```suggestion
        // encryption keys should not be encoded
```

**Review by daira — APPROVED** [review]
Looks good so far.

File: modules/Sources/Models/AddressBookEncryptionKeys.swift (hunk 2 above)
Comment (str4d) [reply modules/Sources/Models/AddressBookEncryptionKeys.swift]: We need to be able to encode the key in some fashion in order to cache it separately from the seed. IDK if this is actually where we want to encode it though, per my other todo.

File: modules/Sources/Generated/XCAssets+Generated.swift [hunk 3]
```diff
@@ -305,6 +306,34 @@ public extension ColorAsset.SystemColor {
   }
 }
 
+public struct DataAsset {
```
Comment (str4d) [inline modules/Sources/Generated/XCAssets+Generated.swift:309]: It's unclear to me what this is for @LukasKorba; it seems auto-generated but I don't see any changes that would have triggered it.

File: modules/Sources/Features/Root/RootInitialization.swift [hunk 4]
```diff
@@ -296,14 +296,16 @@ extension Root {
...
-                        // try walletStorage.importAddressBookEncryptionKeys(keys)
+                        var keys = AddressBookEncryptionKeys.empty
+                        keys.cacheFor(
+                            seed: seedBytes,
+                            account: 0,
+                            network: zcashSDKEnvironment.network
+                        )
+                        // TODO: Why are the address book keys stored in wallet storage?
+                        // Doesn't that mean they require the same authentication as the
+                        // seed phrase, instead of being accessible without biometric auth?
```
Comment (daira) [inline modules/Sources/Features/Root/RootInitialization.swift:307]: There could be situations where you want the history and address book to be behind biometric auth. Maybe this is something we can ask when you add the first address book entry? 

If it's configurable then there should be a single config for the history and the address book, because it's expected that the history will contain addresses that need to be looked up in the address book.

Comment (daira) [reply modules/Sources/Features/Root/RootInitialization.swift:307]: Note that, at least on Android, we have to store the address book encrypted so that it is encrypted when it is backed up. And since the key derivation is from the master seed, there are only two options: 
* require user auth to unlock the address book; or
* store the *derived* key in a place that does not require user auth (we obviously do not want to store the master seed in such a place).

In the latter case, storing the ciphertext and storing the current derived key can't be atomic, because they are in two different places that can't be written together transactionally (I think). Therefore there is necessarily a corner case where the ciphertext doesn't successfully decrypt with the stored key, e.g. when the app is killed after storing one of the ciphertext and the key, and before storing the other. In that case we have to ask for auth to rederive it from the stored salt (which *is* stored atomically with the ciphertext).

Comment (daira) [reply modules/Sources/Features/Root/RootInitialization.swift:307]: Having written that out, I think it might be simplest to require user auth to get the master seed for the derivation now, and defer the implementation of storing the derived key if the user configures it not to require auth to another release. The argument for that is that storing the key isn't required in order for the address book encryption to work, and we would have to handle the case where the correct key isn't stored anyway.

File: modules/Sources/Generated/XCAssets+Generated.swift (hunk 3 above)
Comment (LukasKorba) [reply modules/Sources/Generated/XCAssets+Generated.swift:309]: This is ok, swiftgen has a stencil file that is checking several possible asset types and since project has now a new type of generic data, it must create DataAsset for it, same like ImageAsset, FontAsset, etc.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+extension AddressBookClient {
+    /// Encrypts address book contacts. The structure:
+    ///     [Unencrypted data]    `encryption version`
+    ///     [Unencrypted data]    `salt`
+    ///     [Encrypted data]        `address book version`
+    ///     [Encrypted data]        `timestamp`
+    ///     [Encrypted data]        `contacts`
+    ///
+    /// This method always produces the latest structure with he latest encryption version.
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:23]:
```suggestion
    /// This method always produces the latest structure with the latest
    /// encryption version.
```

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+        @Dependency(\.walletStorage) var walletStorage
+        
+        guard let encryptionKeys = try? walletStorage.exportAddressBookEncryptionKeys(), let addressBookKey = encryptionKeys.getCached(account: 0) else {
+            throw AddressBookClient.AddressBookClientError.missingEncryptionKey
+        }
+        
+        var offset = 0
+        
+        // Deserialize `encryption version`
+        let encryptionVersionBytes = encryptedData.subdata(in: offset..<(offset + MemoryLayout<Int>.size))
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:87]: `subdata(in:)` appears in practice to be memory-unsafe (https://www.programiz.com/online-compiler/0eYmDmico8t92). So for example, if the length of `encryptedData` were less than 4 bytes, this would be unsafe. What I would do in Rust is write a safe helper method that wraps the unsafe one, and only use the helper.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+                if let contact = AddressBookClient.deserializeContact(from: data, at: &offset) {
+                    contacts.append(contact)
+                }
+            }
+            
+            let abContacts = AddressBookContacts(
+                lastUpdated: lastUpdated,
+                version: abVersion,
+                contacts: IdentifiedArrayOf(uniqueElements: contacts)
+            )
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:137]: This duplicates `contactsFrom(unencryptedData: Data)`. It should just call it.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 8]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
+                        let contacts = try AddressBookClient.contactsFrom(encryptedData: contactsData)
+
+                        latestKnownContacts = contacts
+                        return contacts
+                    } else {
+                        // Fallback to the unencrypted file check and resolution
+                        let unencryptedFileURL = documentsDirectory.appendingPathComponent(Constants.unencryptedFilename)
+                        
+                        if let contactsData = try? Data(contentsOf: unencryptedFileURL) {
+                            // file exists, ensure data are parsed, re-saved and file deteled
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:72]:
```suggestion
                            // file exists; ensure data are parsed, re-saved and file deleted
```

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 9]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
 
-                    return decryptedContacts
+                    if let contactsData = try? Data(contentsOf: encryptedFileURL) {
+                        // file exists, try to find the unencrypted file in case the delete failed
+                        let unencryptedFileURL = documentsDirectory.appendingPathComponent(Constants.unencryptedFilename)
+                        if FileManager.default.fileExists(atPath: unencryptedFileURL.path) {
+                            try? FileManager.default.removeItem(at: unencryptedFileURL)
+                        }
+                        
+                        let contacts = try AddressBookClient.contactsFrom(encryptedData: contactsData)
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:63]: This doesn't follow the design:
> If the parse succeeds, also look for a file with the previous filename of the unencrypted address book format and delete it if present.

The attempt to delete the unencrypted file should occur only if the parse succeeds.
```suggestion
                        let contacts = try AddressBookClient.contactsFrom(encryptedData: contactsData)

                        // The encrypted file exists, and can be decrypted and parsed. Try to find the
                        // unencrypted file and delete it, in case an earlier delete failed.
                        let unencryptedFileURL = documentsDirectory.appendingPathComponent(Constants.unencryptedFilename)
                        if FileManager.default.fileExists(atPath: unencryptedFileURL.path) {
                            try? FileManager.default.removeItem(at: unencryptedFileURL)
                        }
```

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 10]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
+                        
+                        if let contactsData = try? Data(contentsOf: unencryptedFileURL) {
+                            // file exists, ensure data are parsed, re-saved and file deteled
+                            let contacts = try AddressBookClient.contactsFrom(unencryptedData: contactsData)
+
+                            // try to encrypt and store the data
+                            do {
+                                try AddressBookClient.storeContacts(contacts, remoteStorage: remoteStorage, remoteStore: false)
+                            } catch {
+                                // the store of the new file failed, skip the file remove
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:79]: This is correct. I updated the design doc to explicitly say that the deletion here should only be done if the store succeeds.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 11]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
+                        // file exists, try to find the unencrypted file in case the delete failed
+                        let unencryptedFileURL = documentsDirectory.appendingPathComponent(Constants.unencryptedFilename)
+                        if FileManager.default.fileExists(atPath: unencryptedFileURL.path) {
+                            try? FileManager.default.removeItem(at: unencryptedFileURL)
+                        }
+                        
+                        let contacts = try AddressBookClient.contactsFrom(encryptedData: contactsData)
+
+                        latestKnownContacts = contacts
+                        return contacts
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:66]: Why do we modify `latestKnownContacts` at every return, rather than having the caller set it?

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 12]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
+                            // file exists, ensure data are parsed, re-saved and file deteled
+                            let contacts = try AddressBookClient.contactsFrom(unencryptedData: contactsData)
+
+                            // try to encrypt and store the data
+                            do {
+                                try AddressBookClient.storeContacts(contacts, remoteStorage: remoteStorage, remoteStore: false)
+                            } catch {
+                                // the store of the new file failed, skip the file remove
+                                latestKnownContacts = contacts
+                                return contacts
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:81]: Shouldn't this be reported as an error?

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 13]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
-        var contacts: [Contact] = []
-        for _ in 0..<contactsCount {
-            if let contact = AddressBookClient.deserializeContact(from: data, at: &offset) {
-                contacts.append(contact)
-            }
-        }
-        
-        let abContacts = AddressBookContacts(
-            lastUpdated: lastUpdated,
+        // Merge strategy
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:167]: This function mixes the merge with the rest of the store logic. I would factor the merge out.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 14]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
-
-        // Serialize `address` (length + UTF-8 bytes)
-        let addressBytes = stringToBytes(contact.id)
-        data.append(contentsOf: intToBytes(addressBytes.count))
-        data.append(contentsOf: addressBytes)
+        remoteContacts.contacts.forEach {
+            var notFound = true
+            var indexToUpdate = -1
+            
+            for i in 0..<syncedContacts.contacts.count {
```
Comment (daira) [inline modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:178]: This is $O(n \cdot n')$ where $n$ is the number of remote contacts and $n'$ is the number of synced contacts. If the contacts are sorted then it should only take $O(m \cdot \log m)$ where $m = \max(n, n')$. Swift probably has a sorted map type with a merge operation. It might be [this one](https://developer.apple.com/documentation/swift/dictionary/merge(_:uniquingkeyswith:)-m2ub).

File: modules/Sources/Features/AddressBook/AddressBookStore.swift [hunk 15]
```diff
@@ -266,21 +266,21 @@ public struct AddressBook {
...
-//                    }
-//                } else {
+                if requestToSync {
+                    return .run { send in
+                        do {
+                            let syncedContacts = try await addressBook.syncContacts(abContacts)
+                            await send(.fetchedABContacts(syncedContacts, false))
+                        } catch {
+                            print("__LD syncContacts Error: \(error.localizedDescription)")
+                            // TODO: FIXME
```
Comment (daira) [inline modules/Sources/Features/AddressBook/AddressBookStore.swift:278]: I know this is an existing TODO, but please file an issue and link it.

File: modules/Sources/Models/AddressBookEncryptionKeys.swift [hunk 16]
```diff
@@ -6,18 +6,103 @@
...
+        guard let info = "file_identifier".data(using: .utf8) else {
+            fatalError("Unable to prepare `file_identifier` info")
+        }
+
+        // Perform HKDF with SHA-256
+        let hkdfKey = HKDF<SHA256>.deriveKey(inputKeyMaterial: key, info: info, outputByteCount: 32)
+        
+        // Convert the HKDF output to a hex string
+        let fileIdentifier = hkdfKey.withUnsafeBytes { rawBytes in
+            rawBytes.map { String(format: "%02x", $0) }.joined()
```
Comment (daira) [inline modules/Sources/Models/AddressBookEncryptionKeys.swift:102]: I documented in the design doc that lowercase hex is used.

File: zashi-internal.entitlements [hunk 17]
```diff
@@ -1,5 +1,22 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
-<dict/>
+<dict>
+	<key>com.apple.developer.associated-domains</key>
+	<array>
+		<string>applinks:secant.flexa.link</string>
+	</array>
```
Comment (daira) [inline zashi-internal.entitlements:8]: Should this be in a different PR?

**Review by daira — CHANGES_REQUESTED** [review]
Changes needed: at least the divergence from the design when attempting to delete the old plaintext file, and error handling for the case where the encrypted file cannot be stored remotely.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 5 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:23]: Updated in code directly.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 7 above)
Comment (daira) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:137]: This is blocking because parsing code especially shouldn't be duplicated.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift [hunk 18]
```diff
@@ -0,0 +1,286 @@
...
+        
+        guard let encryptionKeys = try? walletStorage.exportAddressBookEncryptionKeys(), let addressBookKey = encryptionKeys.getCached(account: 0) else {
+            throw AddressBookClient.AddressBookClientError.missingEncryptionKey
+        }
+        
+        var offset = 0
+        
+        // Deserialize `encryption version`
+        let encryptionVersionBytes = encryptedData.subdata(in: offset..<(offset + MemoryLayout<Int>.size))
+        offset += MemoryLayout<Int>.size
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:88]: If any, Int64 would be the one. Android made it as 8 bytes values, so iOS must follow 8 bytes as well. Int64 can deliver that.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift [hunk 19]
```diff
@@ -0,0 +1,286 @@
...
+        dataForEncryption.append(contentsOf: AddressBookClient.serializeDate(Date()))
+        
+        // Serialize `contacts.count`
+        dataForEncryption.append(contentsOf: intToBytes(abContacts.contacts.count))
+        
+        // Serialize `contacts`
+        abContacts.contacts.forEach { contact in
+            let serializedContact = serializeContact(contact)
+            dataForEncryption.append(serializedContact)
+        }
```
[the remark this answers is not in the document]
Comment (daira) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:49]: Although not strictly blocking, it makes sense to factor this out now. The tests can be deferred.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift [hunk 20]
```diff
@@ -34,304 +39,208 @@ extension AddressBookClient: DependencyKey {
...
+        let fileURL = documentsDirectory.appendingPathComponent(filenameForEncryptedFile)
+        try encryptedContacts.write(to: fileURL)
         
-        // Deserialize `name`
-        guard let name = readString(from: data, at: &offset) else {
-            return nil
+        // store encrypted data to the remote storage
+        if remoteStore {
+            try? remoteStorage.storeAddressBookContacts(encryptedContacts, filenameForEncryptedFile)
         }
```
[the remark this answers is not in the document]
Comment (daira) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:230]: Blocking: report the error.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 11 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:66]: Goal here is to get the local contacts but we don't want to read it from the file every time `allLocalContacts` is called. Instead, it's read only for the first time and next time `latestKnownContacts` delivers the contacts from in-memory. The technology behind this is the dependency in immutable so there must be a subject that can do it for us. `latestKnownContacts` is the mutable subject.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:88]: Or switch to Int32 on both platforms.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 6 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:87]: Resolved with custom

```swift
    private static func subdata(of data: Data, in range: Range<Data.Index>) throws -> Data {
        guard data.count >= range.lowerBound.distance(to: range.upperBound) else {
            throw AddressBookClient.AddressBookClientError.subdataRange
        }
        
        return data.subdata(in: range)
    }
```

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 7 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:137]:
```swift
return try contactsFrom(unencryptedData: data)
```

used inside instead

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 9 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:63]: Swapped the order of functions, attempt to delete is now performed only if parse passed.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 12 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:81]: Throwing an error but leaving the `latestKnownContacts = contacts` so a user can continue and use in-memory contacts.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 13 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:167]: Set for the follow up task #1408

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 14 above)
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:178]: #1408 follow up

File: modules/Sources/Features/AddressBook/AddressBookStore.swift (hunk 15 above)
Comment (LukasKorba) [reply modules/Sources/Features/AddressBook/AddressBookStore.swift:278]: Follow up #1408

File: zashi-internal.entitlements (hunk 17 above)
Comment (LukasKorba) [reply zashi-internal.entitlements:8]: Removed from this PR

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 18 above)
[the remark this answers is not in the document]
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:88]: Updated to MemoryLayout<Int64>.size + moved the value to the constant Constants.int64Size so it's resolved only once and reused.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift (hunk 19 above)
[the remark this answers is not in the document]
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookEncryption.swift:49]: Serizalization code taken out into a separate function.

File: modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift (hunk 20 above)
[the remark this answers is not in the document]
Comment (LukasKorba) [reply modules/Sources/Dependencies/AddressBookClient/AddressBookLiveKey.swift:230]: Re-worked to be able to report `remote store failed` even though local store succeeded.

**Review by LukasKorba — APPROVED** [review]
LGTM
